The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11654669812/job/32448463510) [(plain)](https://github.com/rust-lang/rust/commit/8026d620219da70885e22f0059b920ef00bbc041/checks/32448463510/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.734 Building wheels for collected packages: reuse
#13 2.735   Building wheel for reuse (pyproject.toml): started
#13 2.981   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 2.982   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#13 2.982   Stored in directory: /tmp/pip-ephem-wheel-cache-diznk77e/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 2.985 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.377 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.378 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.900 Collecting virtualenv
#13 3.900 Collecting virtualenv
#13 3.939   Downloading virtualenv-20.27.1-py3-none-any.whl (3.1 MB)
#13 3.997      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 56.2 MB/s eta 0:00:00
#13 4.050 Collecting platformdirs<5,>=3.9.1
#13 4.054   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.089 Collecting filelock<4,>=3.12.2
#13 4.112 Collecting distlib<1,>=0.3.7
#13 4.115   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#13 4.124      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 71.1 MB/s eta 0:00:00
#13 4.124      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 71.1 MB/s eta 0:00:00
#13 4.203 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.377 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.27.1
#13 DONE 4.5s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      171968 kB
DirectMap2M:     7168000 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://static.rust-lang.org/dist/2024-10-16/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/2024-10-16/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt
---
fmt check
Diff in /checkout/compiler/rustc_passes/src/hir_stats.rs:85:
     }
 
     // Record a two-level entry, with a top-level enum type and a variant.
-    fn record_variant<T>(&mut self, label1: &'static str, label2: &'static str, id: Option<HirId>, val: &T) {
+    fn record_variant<T>(
+        label1: &'static str,
+        label2: &'static str,
+        label2: &'static str,
+        id: Option<HirId>,
+        val: &T,
+    ) {
         self.record_inner(label1, Some(label2), id, val);
 
Diff in /checkout/compiler/rustc_passes/src/hir_stats.rs:246:
     }
 
 
     fn visit_foreign_item(&mut self, i: &'v hir::ForeignItem<'v>) {
-        record_variants!(
-            (self, i, i.kind, Some(i.hir_id()), hir, ForeignItem, ForeignItemKind),
-            [Fn, Static, Type]
-        );
+        record_variants!((self, i, i.kind, Some(i.hir_id()), hir, ForeignItem, ForeignItemKind), [
+            Fn, Static, Type
         hir_visit::walk_foreign_item(self, i)
     }
 
Diff in /checkout/compiler/rustc_passes/src/hir_stats.rs:375:
Diff in /checkout/compiler/rustc_passes/src/hir_stats.rs:375:
     }
 
     fn visit_trait_item(&mut self, ti: &'v hir::TraitItem<'v>) {
-        record_variants!(
-            (self, ti, ti.kind, Some(ti.hir_id()), hir, TraitItem, TraitItemKind),
-            [Const, Fn, Type]
-        );
+        record_variants!((self, ti, ti.kind, Some(ti.hir_id()), hir, TraitItem, TraitItemKind), [
+            Const, Fn, Type
         hir_visit::walk_trait_item(self, ti)
     }
 
Diff in /checkout/compiler/rustc_passes/src/hir_stats.rs:388:
Diff in /checkout/compiler/rustc_passes/src/hir_stats.rs:388:
     }
 
     fn visit_impl_item(&mut self, ii: &'v hir::ImplItem<'v>) {
-        record_variants!(
-            (self, ii, ii.kind, Some(ii.hir_id()), hir, ImplItem, ImplItemKind),
-            [Const, Fn, Type]
-        );
+        record_variants!((self, ii, ii.kind, Some(ii.hir_id()), hir, ImplItem, ImplItemKind), [
+            Const, Fn, Type
         hir_visit::walk_impl_item(self, ii)
     }
 
 
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_codegen_gcc/build_system/src/prepare.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/src/clone_gcc.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/src/rust_tools.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/src/main.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/src/rustc_info.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/src/config.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/src/info.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/src/clean.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/src/test.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/src/utils.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/src/fmt.rs" "/checkout/compiler/rustc_passes/src/layout_test.rs" "/checkout/compiler/rustc_passes/src/debugger_visualizer.rs" "/checkout/compiler/rustc_passes/src/reachable.rs" "/checkout/compiler/rustc_passes/src/weak_lang_items.rs" "/checkout/compiler/rustc_passes/src/lib_features.rs" "/checkout/compiler/rustc_passes/src/entry.rs" "/checkout/compiler/rustc_passes/src/hir_id_validator.rs" "/checkout/compiler/rustc_passes/src/liveness.rs" "/checkout/compiler/rustc_passes/src/errors.rs" "/checkout/compiler/rustc_passes/src/loops.rs" "/checkout/compiler/rustc_passes/src/check_const.rs" "/checkout/compiler/rustc_passes/src/upvars.rs" "/checkout/compiler/rustc_passes/src/lang_items.rs" "/checkout/compiler/rustc_passes/src/lib.rs" "/checkout/compiler/rustc_passes/src/stability.rs" "/checkout/compiler/rustc_passes/src/naked_functions.rs" "/checkout/compiler/rustc_passes/src/check_attr.rs" "/checkout/compiler/rustc_passes/src/dead.rs" "/checkout/compiler/rustc_passes/src/diagnostic_items.rs" "/checkout/compiler/rustc_passes/src/liveness/rwu_table.rs" "/checkout/compiler/rustc_passes/src/hir_stats.rs" "/checkout/compiler/rustc_passes/src/abi_test.rs" "/checkout/compiler/rustc_ast_pretty/src/pprust/state/item.rs" "/checkout/compiler/rustc_ast_pretty/src/pprust/state/expr.rs" "/checkout/compiler/rustc_ast_pretty/src/pprust/state/fixup.rs" "/checkout/compiler/rustc_ast_pretty/src/pprust/tests.rs" "/checkout/compiler/rustc_ast_pretty/src/pprust/state.rs" "/checkout/compiler/rustc_ast_pretty/src/pprust/mod.rs" "/checkout/compiler/rustc_ast_pretty/src/pp/convenience.rs" "/checkout/compiler/rustc_ast_pretty/src/pp/ring.rs" "/checkout/compiler/rustc_ast_pretty/src/pp.rs" "/checkout/compiler/rustc_ast_pretty/src/lib.rs" "/checkout/compiler/rustc_ast_pretty/src/helpers.rs" "/checkout/library/std/src/collections/hash/mod.rs" "/checkout/library/std/src/collections/mod.rs" "/checkout/library/std/src/pat.rs" "/checkout/library/std/src/pipe/tests.rs" "/checkout/library/std/src/backtrace/tests.rs" "/checkout/library/std/src/f16.rs" "/checkout/library/std/src/os/xous/services.rs" "/checkout/library/std/src/os/xous/services/net.rs" "/checkout/library/std/src/os/xous/services/dns.rs" "/checkout/library/std/src/os/xous/services/systime.rs" "/checkout/library/std/src/os/xous/services/log.rs" "/checkout/library/std/src/os/xous/services/ticktimer.rs" "/checkout/library/std/src/os/xous/ffi/definitions.rs" "/checkout/library/std/src/os/xous/ffi/definitions/memoryflags.rs" "/checkout/library/std/src/os/xous/ffi.rs" "/checkout/library/std/src/os/xous/mod.rs" "/checkout/library/std/src/os/nto/raw.rs" "/checkout/library/std/src/os/nto/fs.rs" "/checkout/library/std/src/os/nto/mod.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/src/build.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Sun Nov  3 21:13:33 UTC 2024
  network time: Sun, 03 Nov 2024 21:13:33 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/compiler/rustc_passes/src/hir_stats.rs
+++ b/compiler/rustc_passes/src/hir_stats.rs
@@ -3,7 +3,7 @@
 // completely accurate (some things might be counted twice, others missed).
 
 use rustc_ast::visit::BoundKind;
-use rustc_ast::{self as ast, AttrId, NodeId, visit as ast_visit};
+use rustc_ast::{self as ast, NodeId, visit as ast_visit};
 use rustc_data_structures::fx::{FxHashMap, FxHashSet};
 use rustc_hir as hir;
 use rustc_hir::{HirId, intravisit as hir_visit};
@@ -12,13 +12,6 @@ use rustc_middle::ty::TyCtxt;
 use rustc_middle::util::common::to_readable_str;
 use rustc_span::Span;
 use rustc_span::def_id::LocalDefId;
-
-#[derive(Copy, Clone, PartialEq, Eq, Hash)]
-enum Id {
-    Node(HirId),
-    Attr(AttrId),
-    None,
-}
 
 struct NodeStats {
     count: usize,
@@ -62,7 +55,7 @@ impl Node {
 struct StatCollector<'k> {
     krate: Option<Map<'k>>,
     nodes: FxHashMap<&'static str, Node>,
-    seen: FxHashSet<Id>,
+    seen: FxHashSet<HirId>,
 }
 
 pub fn print_hir_stats(tcx: TyCtxt<'_>) {
@@ -87,12 +80,18 @@ pub fn print_ast_stats(krate: &ast::Crate, title: &str, prefix: &str) {
 
 impl<'k> StatCollector<'k> {
     // Record a top-level node.
-    fn record<T>(&mut self, label: &'static str, id: Id, val: &T) {
+    fn record<T>(&mut self, label: &'static str, id: Option<HirId>, val: &T) {
         self.record_inner(label, None, id, val);
     }
 
     // Record a two-level entry, with a top-level enum type and a variant.
-    fn record_variant<T>(&mut self, label1: &'static str, label2: &'static str, id: Id, val: &T) {
+    fn record_variant<T>(
+        &mut self,
+        label1: &'static str,
+        label2: &'static str,
+        id: Option<HirId>,
+        val: &T,
+    ) {
         self.record_inner(label1, Some(label2), id, val);
     }
 
@@ -100,10 +99,10 @@ impl<'k> StatCollector<'k> {
         &mut self,
         label1: &'static str,
         label2: Option<&'static str>,
-        id: Id,
+        id: Option<HirId>,
         val: &T,
     ) {
-        if id != Id::None && !self.seen.insert(id) {
+        if id.is_some_and(|x| !self.seen.insert(x)) {
             return;
         }
 
@@ -191,7 +190,7 @@ macro_rules! record_variants {
 
 impl<'v> hir_visit::Visitor<'v> for StatCollector<'v> {
     fn visit_param(&mut self, param: &'v hir::Param<'v>) {
-        self.record("Param", Id::Node(param.hir_id), param);
+        self.record("Param", Some(param.hir_id), param);
         hir_visit::walk_param(self, param)
     }
 
@@ -221,7 +220,7 @@ impl<'v> hir_visit::Visitor<'v> for StatCollector<'v> {
     }
 
     fn visit_item(&mut self, i: &'v hir::Item<'v>) {
-        record_variants!((self, i, i.kind, Id::Node(i.hir_id()), hir, Item, ItemKind), [
+        record_variants!((self, i, i.kind, Some(i.hir_id()), hir, Item, ItemKind), [
             ExternCrate,
             Use,
             Static,
@@ -243,47 +242,46 @@ impl<'v> hir_visit::Visitor<'v> for StatCollector<'v> {
     }
 
     fn visit_body(&mut self, b: &hir::Body<'v>) {
-        self.record("Body", Id::None, b);
+        self.record("Body", None, b);
         hir_visit::walk_body(self, b);
     }
 
     fn visit_mod(&mut self, m: &'v hir::Mod<'v>, _s: Span, n: HirId) {
-        self.record("Mod", Id::None, m);
+        self.record("Mod", None, m);
         hir_visit::walk_mod(self, m, n)
     }
 
     fn visit_foreign_item(&mut self, i: &'v hir::ForeignItem<'v>) {
-        record_variants!(
-            (self, i, i.kind, Id::Node(i.hir_id()), hir, ForeignItem, ForeignItemKind),
-            [Fn, Static, Type]
-        );
+        record_variants!((self, i, i.kind, Some(i.hir_id()), hir, ForeignItem, ForeignItemKind), [
+            Fn, Static, Type
+        ]);
         hir_visit::walk_foreign_item(self, i)
     }
 
     fn visit_local(&mut self, l: &'v hir::LetStmt<'v>) {
-        self.record("Local", Id::Node(l.hir_id), l);
+        self.record("Local", Some(l.hir_id), l);
         hir_visit::walk_local(self, l)
     }
 
     fn visit_block(&mut self, b: &'v hir::Block<'v>) {
-        self.record("Block", Id::Node(b.hir_id), b);
+        self.record("Block", Some(b.hir_id), b);
         hir_visit::walk_block(self, b)
     }
 
     fn visit_stmt(&mut self, s: &'v hir::Stmt<'v>) {
-        record_variants!((self, s, s.kind, Id::Node(s.hir_id), hir, Stmt, StmtKind), [
+        record_variants!((self, s, s.kind, Some(s.hir_id), hir, Stmt, StmtKind), [
             Let, Item, Expr, Semi
         ]);
         hir_visit::walk_stmt(self, s)
     }
 
     fn visit_arm(&mut self, a: &'v hir::Arm<'v>) {
-        self.record("Arm", Id::Node(a.hir_id), a);
+        self.record("Arm", Some(a.hir_id), a);
         hir_visit::walk_arm(self, a)
     }
 
     fn visit_pat(&mut self, p: &'v hir::Pat<'v>) {
-        record_variants!((self, p, p.kind, Id::Node(p.hir_id), hir, Pat, PatKind), [
+        record_variants!((self, p, p.kind, Some(p.hir_id), hir, Pat, PatKind), [
             Wild,
             Binding,
             Struct,
@@ -304,12 +302,12 @@ impl<'v> hir_visit::Visitor<'v> for StatCollector<'v> {
     }
 
     fn visit_pat_field(&mut self, f: &'v hir::PatField<'v>) {
-        self.record("PatField", Id::Node(f.hir_id), f);
+        self.record("PatField", Some(f.hir_id), f);
         hir_visit::walk_pat_field(self, f)
     }
 
     fn visit_expr(&mut self, e: &'v hir::Expr<'v>) {
-        record_variants!((self, e, e.kind, Id::Node(e.hir_id), hir, Expr, ExprKind), [
+        record_variants!((self, e, e.kind, Some(e.hir_id), hir, Expr, ExprKind), [
             ConstBlock, Array, Call, MethodCall, Tup, Binary, Unary, Lit, Cast, Type, DropTemps,
             Let, If, Loop, Match, Closure, Block, Assign, AssignOp, Field, Index, Path, AddrOf,
             Break, Continue, Ret, Become, InlineAsm, OffsetOf, Struct, Repeat, Yield, Err
@@ -318,12 +316,12 @@ impl<'v> hir_visit::Visitor<'v> for StatCollector<'v> {
     }
 
     fn visit_expr_field(&mut self, f: &'v hir::ExprField<'v>) {
-        self.record("ExprField", Id::Node(f.hir_id), f);
+        self.record("ExprField", Some(f.hir_id), f);
         hir_visit::walk_expr_field(self, f)
     }
 
     fn visit_ty(&mut self, t: &'v hir::Ty<'v>) {
-        record_variants!((self, t, t.kind, Id::Node(t.hir_id), hir, Ty, TyKind), [
+        record_variants!((self, t, t.kind, Some(t.hir_id), hir, Ty, TyKind), [
             InferDelegation,
             Slice,
             Array,
@@ -345,17 +343,17 @@ impl<'v> hir_visit::Visitor<'v> for StatCollector<'v> {
     }
 
     fn visit_generic_param(&mut self, p: &'v hir::GenericParam<'v>) {
-        self.record("GenericParam", Id::Node(p.hir_id), p);
+        self.record("GenericParam", Some(p.hir_id), p);
         hir_visit::walk_generic_param(self, p)
     }
 
     fn visit_generics(&mut self, g: &'v hir::Generics<'v>) {
-        self.record("Generics", Id::None, g);
+        self.record("Generics", None, g);
         hir_visit::walk_generics(self, g)
     }
 
     fn visit_where_predicate(&mut self, p: &'v hir::WherePredicate<'v>) {
-        record_variants!((self, p, p, Id::None, hir, WherePredicate, WherePredicate), [
+        record_variants!((self, p, p, None, hir, WherePredicate, WherePredicate), [
             BoundPredicate,
             RegionPredicate,
             EqPredicate
@@ -371,66 +369,64 @@ impl<'v> hir_visit::Visitor<'v> for StatCollector<'v> {
         _: Span,
         id: LocalDefId,
     ) {
-        self.record("FnDecl", Id::None, fd);
+        self.record("FnDecl", None, fd);
         hir_visit::walk_fn(self, fk, fd, b, id)
     }
 
     fn visit_use(&mut self, p: &'v hir::UsePath<'v>, hir_id: HirId) {
         // This is `visit_use`, but the type is `Path` so record it that way.
-        self.record("Path", Id::None, p);
+        self.record("Path", None, p);
         hir_visit::walk_use(self, p, hir_id)
     }
 
     fn visit_trait_item(&mut self, ti: &'v hir::TraitItem<'v>) {
-        record_variants!(
-            (self, ti, ti.kind, Id::Node(ti.hir_id()), hir, TraitItem, TraitItemKind),
-            [Const, Fn, Type]
-        );
+        record_variants!((self, ti, ti.kind, Some(ti.hir_id()), hir, TraitItem, TraitItemKind), [
+            Const, Fn, Type
+        ]);
         hir_visit::walk_trait_item(self, ti)
     }
 
     fn visit_trait_item_ref(&mut self, ti: &'v hir::TraitItemRef) {
-        self.record("TraitItemRef", Id::Node(ti.id.hir_id()), ti);
+        self.record("TraitItemRef", Some(ti.id.hir_id()), ti);
         hir_visit::walk_trait_item_ref(self, ti)
     }
 
     fn visit_impl_item(&mut self, ii: &'v hir::ImplItem<'v>) {
-        record_variants!(
-            (self, ii, ii.kind, Id::Node(ii.hir_id()), hir, ImplItem, ImplItemKind),
-            [Const, Fn, Type]
-        );
+        record_variants!((self, ii, ii.kind, Some(ii.hir_id()), hir, ImplItem, ImplItemKind), [
+            Const, Fn, Type
+        ]);
         hir_visit::walk_impl_item(self, ii)
     }
 
     fn visit_foreign_item_ref(&mut self, fi: &'v hir::ForeignItemRef) {
-        self.record("ForeignItemRef", Id::Node(fi.id.hir_id()), fi);
+        self.record("ForeignItemRef", Some(fi.id.hir_id()), fi);
         hir_visit::walk_foreign_item_ref(self, fi)
     }
 
     fn visit_impl_item_ref(&mut self, ii: &'v hir::ImplItemRef) {
-        self.record("ImplItemRef", Id::Node(ii.id.hir_id()), ii);
+        self.record("ImplItemRef", Some(ii.id.hir_id()), ii);
         hir_visit::walk_impl_item_ref(self, ii)
     }
 
     fn visit_param_bound(&mut self, b: &'v hir::GenericBound<'v>) {
-        record_variants!((self, b, b, Id::None, hir, GenericBound, GenericBound), [
+        record_variants!((self, b, b, None, hir, GenericBound, GenericBound), [
             Trait, Outlives, Use
         ]);
         hir_visit::walk_param_bound(self, b)
     }
 
     fn visit_field_def(&mut self, s: &'v hir::FieldDef<'v>) {
-        self.record("FieldDef", Id::Node(s.hir_id), s);
+        self.record("FieldDef", Some(s.hir_id), s);
         hir_visit::walk_field_def(self, s)
     }
 
     fn visit_variant(&mut self, v: &'v hir::Variant<'v>) {
-        self.record("Variant", Id::None, v);
+        self.record("Variant", None, v);
         hir_visit::walk_variant(self, v)
     }
 
     fn visit_generic_arg(&mut self, ga: &'v hir::GenericArg<'v>) {
-        record_variants!((self, ga, ga, Id::Node(ga.hir_id()), hir, GenericArg, GenericArg), [
+        record_variants!((self, ga, ga, Some(ga.hir_id()), hir, GenericArg, GenericArg), [
             Lifetime, Type, Const, Infer
         ]);
         match ga {
@@ -442,50 +438,50 @@ impl<'v> hir_visit::Visitor<'v> for StatCollector<'v> {
     }
 
     fn visit_lifetime(&mut self, lifetime: &'v hir::Lifetime) {
-        self.record("Lifetime", Id::Node(lifetime.hir_id), lifetime);
+        self.record("Lifetime", Some(lifetime.hir_id), lifetime);
         hir_visit::walk_lifetime(self, lifetime)
     }
 
     fn visit_path(&mut self, path: &hir::Path<'v>, _id: HirId) {
-        self.record("Path", Id::None, path);
+        self.record("Path", None, path);
         hir_visit::walk_path(self, path)
     }
 
     fn visit_path_segment(&mut self, path_segment: &'v hir::PathSegment<'v>) {
-        self.record("PathSegment", Id::None, path_segment);
+        self.record("PathSegment", None, path_segment);
         hir_visit::walk_path_segment(self, path_segment)
     }
 
     fn visit_generic_args(&mut self, ga: &'v hir::GenericArgs<'v>) {
-        self.record("GenericArgs", Id::None, ga);
+        self.record("GenericArgs", None, ga);
         hir_visit::walk_generic_args(self, ga)
     }
 
     fn visit_assoc_item_constraint(&mut self, constraint: &'v hir::AssocItemConstraint<'v>) {
-        self.record("AssocItemConstraint", Id::Node(constraint.hir_id), constraint);
+        self.record("AssocItemConstraint", Some(constraint.hir_id), constraint);
         hir_visit::walk_assoc_item_constraint(self, constraint)
     }
 
     fn visit_attribute(&mut self, attr: &'v ast::Attribute) {
-        self.record("Attribute", Id::Attr(attr.id), attr);
+        self.record("Attribute", None, attr);
     }
 
     fn visit_inline_asm(&mut self, asm: &'v hir::InlineAsm<'v>, id: HirId) {
-        self.record("InlineAsm", Id::None, asm);
+        self.record("InlineAsm", None, asm);
         hir_visit::walk_inline_asm(self, asm, id);
     }
 }
 
 impl<'v> ast_visit::Visitor<'v> for StatCollector<'v> {
     fn visit_foreign_item(&mut self, i: &'v ast::ForeignItem) {
-        record_variants!((self, i, i.kind, Id::None, ast, ForeignItem, ForeignItemKind), [
+        record_variants!((self, i, i.kind, None, ast, ForeignItem, ForeignItemKind), [
             Static, Fn, TyAlias, MacCall
         ]);
         ast_visit::walk_item(self, i)
     }
 
     fn visit_item(&mut self, i: &'v ast::Item) {
-        record_variants!((self, i, i.kind, Id::None, ast, Item, ItemKind), [
+        record_variants!((self, i, i.kind, None, ast, Item, ItemKind), [
             ExternCrate,
             Use,
             Static,
@@ -510,34 +506,34 @@ impl<'v> ast_visit::Visitor<'v> for StatCollector<'v> {
     }
 
     fn visit_local(&mut self, l: &'v ast::Local) {
-        self.record("Local", Id::None, l);
+        self.record("Local", None, l);
         ast_visit::walk_local(self, l)
     }
 
     fn visit_block(&mut self, b: &'v ast::Block) {
-        self.record("Block", Id::None, b);
+        self.record("Block", None, b);
         ast_visit::walk_block(self, b)
     }
 
     fn visit_stmt(&mut self, s: &'v ast::Stmt) {
-        record_variants!((self, s, s.kind, Id::None, ast, Stmt, StmtKind), [
+        record_variants!((self, s, s.kind, None, ast, Stmt, StmtKind), [
             Let, Item, Expr, Semi, Empty, MacCall
         ]);
         ast_visit::walk_stmt(self, s)
     }
 
     fn visit_param(&mut self, p: &'v ast::Param) {
-        self.record("Param", Id::None, p);
+        self.record("Param", None, p);
         ast_visit::walk_param(self, p)
     }
 
     fn visit_arm(&mut self, a: &'v ast::Arm) {
-        self.record("Arm", Id::None, a);
+        self.record("Arm", None, a);
         ast_visit::walk_arm(self, a)
     }
 
     fn visit_pat(&mut self, p: &'v ast::Pat) {
-        record_variants!((self, p, p.kind, Id::None, ast, Pat, PatKind), [
+        record_variants!((self, p, p.kind, None, ast, Pat, PatKind), [
             Wild,
             Ident,
             Struct,
@@ -563,7 +559,7 @@ impl<'v> ast_visit::Visitor<'v> for StatCollector<'v> {
     fn visit_expr(&mut self, e: &'v ast::Expr) {
         #[rustfmt::skip]
         record_variants!(
-            (self, e, e.kind, Id::None, ast, Expr, ExprKind),
+            (self, e, e.kind, None, ast, Expr, ExprKind),
             [
                 Array, ConstBlock, Call, MethodCall, Tup, Binary, Unary, Lit, Cast, Type, Let,
                 If, While, ForLoop, Loop, Match, Closure, Block, Await, TryBlock, Assign,
@@ -576,7 +572,7 @@ impl<'v> ast_visit::Visitor<'v> for StatCollector<'v> {
     }
 
     fn visit_ty(&mut self, t: &'v ast::Ty) {
-        record_variants!((self, t, t.kind, Id::None, ast, Ty, TyKind), [
+        record_variants!((self, t, t.kind, None, ast, Ty, TyKind), [
             Slice,
             Array,
             Ptr,
@@ -603,12 +599,12 @@ impl<'v> ast_visit::Visitor<'v> for StatCollector<'v> {
     }
 
     fn visit_generic_param(&mut self, g: &'v ast::GenericParam) {
-        self.record("GenericParam", Id::None, g);
+        self.record("GenericParam", None, g);
         ast_visit::walk_generic_param(self, g)
     }
 
     fn visit_where_predicate(&mut self, p: &'v ast::WherePredicate) {
-        record_variants!((self, p, p, Id::None, ast, WherePredicate, WherePredicate), [
+        record_variants!((self, p, p, None, ast, WherePredicate, WherePredicate), [
             BoundPredicate,
             RegionPredicate,
             EqPredicate
@@ -617,12 +613,12 @@ impl<'v> ast_visit::Visitor<'v> for StatCollector<'v> {
     }
 
     fn visit_fn(&mut self, fk: ast_visit::FnKind<'v>, _: Span, _: NodeId) {
-        self.record("FnDecl", Id::None, fk.decl());
+        self.record("FnDecl", None, fk.decl());
         ast_visit::walk_fn(self, fk)
     }
 
     fn visit_assoc_item(&mut self, i: &'v ast::AssocItem, ctxt: ast_visit::AssocCtxt) {
-        record_variants!((self, i, i.kind, Id::None, ast, AssocItem, AssocItemKind), [
+        record_variants!((self, i, i.kind, None, ast, AssocItem, AssocItemKind), [
             Const,
             Fn,
             Type,
@@ -634,19 +630,19 @@ impl<'v> ast_visit::Visitor<'v> for StatCollector<'v> {
     }
 
     fn visit_param_bound(&mut self, b: &'v ast::GenericBound, _ctxt: BoundKind) {
-        record_variants!((self, b, b, Id::None, ast, GenericBound, GenericBound), [
+        record_variants!((self, b, b, None, ast, GenericBound, GenericBound), [
             Trait, Outlives, Use
         ]);
         ast_visit::walk_param_bound(self, b)
     }
 
     fn visit_field_def(&mut self, s: &'v ast::FieldDef) {
-        self.record("FieldDef", Id::None, s);
+        self.record("FieldDef", None, s);
         ast_visit::walk_field_def(self, s)
     }
 
     fn visit_variant(&mut self, v: &'v ast::Variant) {
-        self.record("Variant", Id::None, v);
+        self.record("Variant", None, v);
         ast_visit::walk_variant(self, v)
     }
 
@@ -660,7 +656,7 @@ impl<'v> ast_visit::Visitor<'v> for StatCollector<'v> {
     // common than the former case, so we implement this visitor and tolerate
     // the double counting in the former case.
     fn visit_path_segment(&mut self, path_segment: &'v ast::PathSegment) {
-        self.record("PathSegment", Id::None, path_segment);
+        self.record("PathSegment", None, path_segment);
         ast_visit::walk_path_segment(self, path_segment)
     }
 
@@ -669,7 +665,7 @@ impl<'v> ast_visit::Visitor<'v> for StatCollector<'v> {
     // common, so we implement `visit_generic_args` and tolerate the double
     // counting in the former case.
     fn visit_generic_args(&mut self, g: &'v ast::GenericArgs) {
-        record_variants!((self, g, g, Id::None, ast, GenericArgs, GenericArgs), [
+        record_variants!((self, g, g, None, ast, GenericArgs, GenericArgs), [
             AngleBracketed,
             Parenthesized,
             ParenthesizedElided
@@ -678,24 +674,24 @@ impl<'v> ast_visit::Visitor<'v> for StatCollector<'v> {
     }
 
     fn visit_attribute(&mut self, attr: &'v ast::Attribute) {
-        record_variants!((self, attr, attr.kind, Id::None, ast, Attribute, AttrKind), [
+        record_variants!((self, attr, attr.kind, None, ast, Attribute, AttrKind), [
             Normal, DocComment
         ]);
         ast_visit::walk_attribute(self, attr)
     }
 
     fn visit_expr_field(&mut self, f: &'v ast::ExprField) {
-        self.record("ExprField", Id::None, f);
+        self.record("ExprField", None, f);
         ast_visit::walk_expr_field(self, f)
     }
 
     fn visit_crate(&mut self, krate: &'v ast::Crate) {
-        self.record("Crate", Id::None, krate);
+        self.record("Crate", None, krate);
         ast_visit::walk_crate(self, krate)
     }
 
     fn visit_inline_asm(&mut self, asm: &'v ast::InlineAsm) {
-        self.record("InlineAsm", Id::None, asm);
+        self.record("InlineAsm", None, asm);
         ast_visit::walk_inline_asm(self, asm)
     }
 }

